### PR TITLE
Fix for KEYCLOAK-18914 - Base64 decoding

### LINKED ISSUE #9382: Password credential decoding from DB may fail in rare cases - No login possible
### Describe the bug

See https://issues.redhat.com/browse/KEYCLOAK-18914 for the details

### Version

12.0.0, 13.0.0, 14.0.0, 15.0.0, 16.0.0

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by hc42**
Fixed https://github.com/keycloak/keycloak/issues/9382

- Do not try to gzunzip based on input stream content as this may lead to unexpected behaviour in some cases were the data is not gzip encoded (e.g. random input as described in the but ticket). Use option flags instead for explicit control.
- Unify two parallel exiting implementations of the Base64 class (they already started diverging in maintenance).

### CODE REVIEW

**Review by mposolda — CHANGES_REQUESTED** [review]
@hc42 Thanks for the fix! The fix itself looks good to me as long as all the tests will pass.

Just one small thing: We are moving from JIRA to GH issues. Is it possible if you create new GH issue for this (It should be ok to add just small description as "See https://issues.redhat.com/browse/KEYCLOAK-18914 for the details") and change the commit message in this PR to link to this GH issue? See https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md for the details.

**Comment by stianst** [comment]
Java has had Base64 support built-in since 1.8, and we do not support any older versions any longer. So would it not be better to just switch to the built-in support in Java?

https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Decoder.html

**Comment by hmlnarik** [comment]
Base64 is used in SAML adapters. Have we ceased supporting EAP 6.4 adapters already?

**Comment by stianst** [comment]
> Base64 is used in SAML adapters. Have we ceased supporting EAP 6.4 adapters already?

No, but EAP 6.4 surely isn't supported on older Java than 1.8?

**Comment by hmlnarik** [comment]
> > Base64 is used in SAML adapters. Have we ceased supporting EAP 6.4 adapters already?
> 
> No, but EAP 6.4 surely isn't supported on older Java than 1.8?

It is [1] supported until 2025 [2].

I would also prefer removing the need to keep jdk 7 classes as it became a blocker for many features. However for this PR I'd rather keep Base64 implementation in the codebase and treat this issue separately.

[1] https://access.redhat.com/articles/111663
[2] https://access.redhat.com/support/policy/updates/jboss_notes/#p_eap

**Comment by hc42** [comment]
Hi,
I updated the commit message with the new Github issue as proposed by @mposolda . I also rebased to the current main branch (commit 208e45c).

**Comment by stianst** [comment]
> > > Base64 is used in SAML adapters. Have we ceased supporting EAP 6.4 adapters already?
> > 
> > No, but EAP 6.4 surely isn't supported on older Java than 1.8?
> 
> It is [1] supported until 2025 [2].
> 
> I would also prefer removing the need to keep jdk 7 classes as it became a blocker for many features. However for this PR I'd rather keep Base64 implementation in the codebase and treat this issue separately.
> 
> [1] https://access.redhat.com/articles/111663
> [2] https://access.redhat.com/support/policy/updates/jboss_notes/#p_eap
> 
> 

Sure, 6.4 is supported until 2025, but Java 1.7 hasn't been supported for over a year: https://access.redhat.com/articles/1299013

**Review by mposolda — APPROVED** [review]
@stianst @hmlnarik Ok, I am merging this.

@hc42 Thanks for the clarification regarding commit message.

File: common/src/main/java/org/keycloak/common/util/Base64.java [hunk 1]
```diff
@@ -733,7 +738,7 @@ public static String encodeBytes( byte[] source ) {
      * Example options:<pre>
      *   GZIP: gzip-compresses object before encoding it.
      *   DO_BREAK_LINES: break lines at 76 characters
-     *     <i>Note: Technically, this makes your encoding non-compliant.</i>
+     *     <i>Note: Technically, without line break your encoding may become non-compliant (see rfc2045 and rfc4648).</i>
```
Comment (hshort) [inline common/src/main/java/org/keycloak/common/util/Base64.java:741]: Unfortunately adding line breaks causes many SAML Service Providers to fail validating signatures. This is a blocking issue for people being able to upgrade Keycloak now.

**Pull request merged**